contrib/zenith/pagestore_prefetch.c

```diff
@@ -0,0 +1,440 @@
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+#include "miscadmin.h"
+#include "pgstat.h"
+#include "pagestore_client.h"
+#include "postmaster/bgworker.h"
+#include "storage/relfilenode.h"
+#include "storage/buf_internals.h"
+#include "storage/latch.h"
+#include "storage/ipc.h"
+#include "storage/lwlock.h"
+#include "utils/dynahash.h"
+#include "utils/guc.h"
+#include "utils/timestamp.h"
+#include "pagestore_client.h"
+
+/*
+ * Prefetch is using fixed size ring buffer and postgres hash table. Both are located in shared memory.
+ * Backends are adding new entries to the ring buffer and prefetch background worker is taken them from ring buffer.
+ * When smgrprefetch() is called, we try to locate entry in hash table. If it is not present yet, then we add entry to ring buffer.
+ * We do no check overflow of ring buffer: if entry was not yet proceeded by prefetch worker, then we just replace it because
+ * this request is considered as deteriorated.
+ *
+ * When smgrread is called, we first try to locate entry in prefetch hash table. If request is present in hash table, then it can be in
+ * one of three states:
+ * 1. Not yet proceeded (sent to page server)
+ * 2. Sent to page server, but response is not yet received.
+ * 3. Proceeded: prefetched page is placed in ring buffer
+ * In the first case we just cancel prefetch request and let backend request page from page server itself.
+ * In the second case we wait until request is completed.
+ * And in the third case copy data from ring buffer to the destination and immediately return
+ */
+
+typedef enum {
+	QUEUED,
+	IN_PROGRESS,
+	COMPLETED,
+	CANCELED,
+} PrefetchState;
+
+typedef struct PrefetchEntry {
+	BufferTag     tag;
+	XLogRecPtr    lsn;
+	uint32        index;  /* index of entry in prefetch ring buffer */
+	PrefetchState state;
+	Latch*        waiting_backend; /* latch to be signaled when block is fetched */
+} PrefetchEntry;
+
+typedef struct PrefetchControl {
+	size_t curr; /* position in ring buffer */
+	Latch* waiting_prefetcher; /* latch to wakeup prefetcher */
+	PrefetchEntry* entries[FLEXIBLE_ARRAY_MEMBER]; /* ring buffer with size == prefetch_buffer_size */
+} PrefetchControl;
+
+static int prefetch_buffer_size;
+static PrefetchControl* prefetch_control;
+static char* prefetch_buffer;
+static HTAB* prefetch_hash;
+static LWLockId prefetch_lock;
+static shmem_startup_hook_type prev_shmem_startup_hook = NULL;
+static volatile bool prefetch_cancel;
+
+size_t n_prefetch_requests;
+size_t n_prefetch_hits;
+size_t max_merged_prefetch_requests;
+
+static void
+zenith_prefetch_shmem_startup(void)
+{
+	bool found;
+
+	if (prev_shmem_startup_hook) {
+		prev_shmem_startup_hook();
+    }
+
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+	prefetch_buffer = ShmemInitStruct("zenith_prefetch",
+									  sizeof(PrefetchControl) +
+									  (BLCKSZ+sizeof(PrefetchEntry*))*prefetch_buffer_size,
```

Comment:
This subexpression can be found in `zenith_shmem_size`. Could you try extracting this one instead?